packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js

```diff
@@ -0,0 +1,175 @@
+    if (this.denominatorPriceFeed) {
+      this.allPriceFeeds = this.allPriceFeeds.concat(this.denominatorPriceFeed);
+    }
+
+    // Helper modules.
+    this.web3 = web3;
+    this.toBN = this.web3.utils.toBN;
+    this.toWei = this.web3.utils.toWei;
+    this.logger = logger;
+
+    // The precision that the user wants to return prices in must match all basket constituent price feeds and the denominator.
+    this.decimals = this.allPriceFeeds[0].getPriceFeedDecimals();
+
+    // Scale `number` by 10**decimals.
+    this.convertPriceFeedDecimals = number => {
+      // Converts price result to wei
+      // returns price conversion to correct decimals as a big number
+      return this.toBN(parseFixed(number.toString(), this.decimals).toString());
+    };
+  }
+
+  // Given lists of experimental and baseline prices, and a denominator price,
+  // return the spread price, which is:
+  // (avg(experimental) - avg(baseline) + 1) / denominator
+  _getSpreadFromBasketPrices(experimentalPrices, baselinePrices, denominatorPrice) {
+    // Compute experimental basket mean.
+    if (
+      experimentalPrices.length === 0 ||
+      experimentalPrices.some(element => element === undefined || element === null)
+    ) {
+      return null;
+    }
+    const experimentalMean = this._computeMean(experimentalPrices);
+    this.logger.debug({
+      at: "BasketSpreadPriceFeed",
+      message: "Average of experimental prices",
+      mean: experimentalMean.toString()
+    });
+
+    // Second, compute the average of the baseline pricefeeds.
+    if (baselinePrices.length === 0 || baselinePrices.some(element => element === undefined || element === null)) {
+      return null;
+    }
+    const baselineMean = this._computeMean(baselinePrices);
+    this.logger.debug({
+      at: "BasketSpreadPriceFeed",
+      message: "Average of baseline prices",
+      mean: baselineMean.toString()
+    });
+
+    // All calculations within this if statement will produce unexpected results if any of the
+    // experimental mean, baseline mean, or denominator price are NOT in the same precision as
+    // the one that this.convertPriceFeedDecimals() uses.
+    if (!baselineMean || !experimentalMean) return null;
+
+    // TODO: Parameterize the lower (0) and upper (2) bounds, as well as allow for custom "spreadValue" formulas,
+    // for example we might not want to have the spread centered around 1, like it is here:
+    let spreadValue = experimentalMean.sub(baselineMean).add(this.convertPriceFeedDecimals("1"));
+    this.logger.debug({
+      at: "BasketSpreadPriceFeed",
+      message: "Basket spread value",
+      spreadValue: spreadValue.toString()
+    });
+
+    // Ensure non-negativity
+    if (spreadValue.lt(this.toBN("0"))) {
+      spreadValue = this.toBN("0");
+    }
+    // Ensure symmetry
+    else if (spreadValue.gt(this.convertPriceFeedDecimals("2"))) {
+      spreadValue = this.convertPriceFeedDecimals("2");
```

Comment:
What is the rationale for these bounds?

Reply:
As defined in UMIP 28: https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-28.md#implementation